Pointwise Tag, use in aot_autograd partitioner

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90029

Takes the pointwise op list from [DTensor](https://github.com/pytorch/pytorch/blob/master/torch/distributed/_tensor/ops/pointwise_ops.py#L36) as an initially starting point for pointwise ops, and feeds them to the aot autograd partitioner. 

Edit: expanded to cover all ops in Unary, Binary, Ternary, and TensorCompare.

eellison committed:
Add Pointwise Tag from pointwise set in DTensor, use in aot_autograd partitioner

[ghstack-poisoned]

eellison commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -40,3 +40,8 @@
           type promotion and boardcasting ops.
           Canonical aten ops is also effectively the opset produced by torchdynamo.export(aten_graph=True),
           and thus can be used as an opset for export purpose.
+- tag: pointwise
```
CC @wanchaol, is there way to remove https://github.com/pytorch/pytorch/blob/master/torch/distributed/_tensor/ops/pointwise_ops.py#L36 and use the tag ? 

I see elsewhere there is `Below ops are some pointwise ops that are yet to be supported` so I guess you need to differentiate between supported pointwise op and unsupported pointwise ops

wanchaol replied:
This is nice! Thanks for adding the tag! I have been thinking about tagging a while ago and wonder if DTensor could just simply utilize the tags. I think it should be doable, I'll try it out once you land the diff, thanks!

wanchaol commented on functorch/_src/partitioners.py:
```diff
@@ -346,6 +365,9 @@ def is_tensor_node(x):
         default_recomputable_ops += [aten.index]
 
+        # add more generally ?
+        default_recomputable_ops += pointwise_ops()
```
I recalled the DTensor pointwise op set might not be a complete set of all the pointwise operators in ATen (although it covers most of them). Do you need to include all possible pointwise ops initially, or we can increase the coverage step by step?

ezyang replied:
Does dtensor need the narrower set here?

wanchaol replied:
Yeah possibly DTensor needs a narrower set as some of the pointwise op might not make too much sense to DTensor, but I would prefer the tag in native_functions to reflect all pointwise ops, and leave the subsystems (i.e. DTensor or aot autograd) to decide and filter out the ops needed from the tagged set.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -244,14 +244,15 @@
     CUDA: native_dropout_cuda
     NestedTensorCPU, NestedTensorCUDA: native_dropout_nested
-  tags: nondeterministic_seeded, canonical
+  tags: [nondeterministic_seeded, canonical]
```
Hmm. This is OK but I am fundamentally mistrustful of leaning too much on YAML's built in parser; splitting on the comma was not too hard to do either haha

eellison replied:
Happy to do either but this seemed more in precedent of how we split everything else and has better highlighting in a code editor. Btw I had a small script to edit yaml files that was helpful here I can share.

ezyang commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -40,3 +40,8 @@
+  desc: |
+          Pointwise operators are operators where each element of the output is computed only by accessing
+          the corresponding element of all the broadcasted inputs. The output shape will be the broadcasted
+          shape of the inputs.
```
Does pointwise imply that it is implemented with TensorIterator?

eellison replied:
Pointwise is more a description of how an operator is computed than its underlying implementation. I would imagine all pointwise operators should be computed with TensorIterator but I don't think that's a necessary condition. Open to input though.

ezyang replied:
The main reason I ask is because there is a bunch of subtle striding behavior, which is unlikely to be done correctly if you're not using TensorIterator under the hood. So, it matters materially for the more subtle invariants whether or not you're buying into "all of the TensorIterator semantics". We don't have to call it TI but I am interested in knowing if we are giving these guarantees.

eellison replied:
I don't think so. although would be interested to hear from @wanchaol and @Chillee, who have essentially been maintaining their own pointwise op set.

wanchaol replied:
To add some context about the DTensor pointwise op set, it was initially added from manual inspection of every op in the native_functions.yaml by referencing https://pytorch.org/docs/stable/torch.html#pointwise-ops, then added some missing pointwise ops when we actually tried on real models. I believe this set is still not completely including all possible pointwise ops yet (but it's close).

For the meaning of pointwise ops, I think the description from @eellison should be a fair enough description of the definition of a pointwise op, but I agree we might need some formal algorithmic guard on it to ensure existing and newly added pointwise op get tagged too.

ezyang commented on functorch/_src/partitioners.py:
```diff
@@ -246,6 +247,24 @@ def _count_ops(graph):
+                ops.append(opoverloadpacket)
+                break
+
+    return ops
```
If at all possible, we should figure out a way to do this that doesn't involve glomming all of the operators into an LRU cache. The main hazard to doing it this way is that torch.ops.aten is lazily loaded, so there isn't a guarantee that you will actually have all of the operators at the time you call this function. And this will be even worse if you want to support pointwise ops outside of the aten namespace

anjali411 replied:
yeah that's a good idea. I think we can just create it in codegen when the native functions are parsed. 

Currently, there's no API for users to add tags from Python Library API but in the future, we'll have to ensure that this db is up-to-date when new ops are added.

ezyang replied:
I mentioned this in WC but to repeat it here: for use in the partitioner, you don't need the entire set, at the point where you need to test if an op is pointwise or not, just check the tag directly.

ezyang commented:
On my end, I need a more clarity on what exactly pointwise means, and when it is permissible to add an operator to this set. "We just used the same list from DTensor" does it cut it in the long term.

eellison commented:
@ezyang 

I took from the DTensor supported list initially so that it would be easier to remove that list moving forward if there were extra invariants in that list that are separate from the definition of pointwise ops, and so that you wouldn't have to individually review every operator I'm adding here. 

I think :
```
Pointwise operators are operators where each element of the output is computed only by accessing
the corresponding element of all the broadcasted inputs. The output shape will be the broadcasted
shape of the inputs.
``` 
is a concise and adequate representation of pointwise. I'm happy to hear alternative definitions. 

As it stands, the list in AOT Autograd was missing the following operators, which prevents efficient recomputation and subsequent speeding up of dynamo models.

```
aten::acosh
aten::addcdiv
aten::angle
aten::asinh
aten::bitwise_left_shift
aten::bitwise_right_shift
aten::bitwise_xor
aten::clip
aten::conj_physical
aten::deg2rad
aten::digamma
aten::erfinv
aten::exp2
aten::hypot
aten::i0
aten::igamma
aten::igammac
aten::logaddexp
aten::logaddexp2
aten::logical_not
aten::logical_xor
aten::logit
aten::masked_fill
aten::mvlgamma
aten::nan_to_num
aten::native_dropout_backward
aten::nextafter
aten::positive
aten::sgn
aten::sigmoid_backward
aten::signbit
aten::sinc
aten::square
aten::tanh_backward
aten::true_divide
aten::xlogy
```

ezyang commented:
I guess my main concern is (1) how do we know this list as is, is correct, and (2) how do we make sure people keep this list up to date as new operators are added. In the original tags proposal, every tag was supposed to be accompanied with a test that could programmatically determine if an operator was in the set or not. There isn't any such test currently, which suggests that we're pretty unlikely to keep this list up to date going forward

wanchaol commented:
> I guess my main concern is (1) how do we know this list as is, is correct, and (2) how do we make sure people keep this list up to date as new operators are added. In the original tags proposal, every tag was supposed to be accompanied with a test that could programmatically determine if an operator was in the set or not. There isn't any such test currently, which suggests that we're pretty unlikely to keep this list up to date going forward

To ensure that the list is correct and keep the list up to date, I recalled I had a conversation with @albanD about this. Alban suggested that we can compute the vjp of the operator, and if the gradient is a diagonal matrix, then we are certain that this is a pointwise op. Maybe we can use this approach to guard the list?

ezyang commented:
Checking that the vjp is diagonal will certainly let you test that there aren't functions that are incorrectly tagged this way. But do you really want to test the inverse (that a new function isn't missing the tag) this way? Seems a bit questionable.

albanD commented:
tbh it is pretty hard to test  programmatically here with the definition above. You will have to test every function, every sample for that function and every input value for these.
So whatever test we want to use (Jacobian, or eps difference on the input, etc) will be very expensive.

I think there are two options here:
- Change the definition to be something not about the values of the function but the function itself. Such that we can check it without having to test the input/output mapping.
- Add a new periodic validate_tags test that is going to be very very expensive but we don't run it often.

ezyang commented:
I'm also ok with white box approaches; e.g., we grep the C++ source code or something and check it users TensorIterator, for example

eellison commented:
@wanchaol how did you come up with this list originally ?

eellison committed:
Update on "Add Pointwise Tag from pointwise set in DTensor, use in aot_autograd partitioner"


Takes the pointwise op list from [DTensor](https://github.com/pytorch/pytorch/blob/master/torch/distributed/_tensor/ops/pointwise_ops.py#L36) as an initially starting point for pointwise ops, and feeds them to the aot autograd partitioner. 


[ghstack-poisoned]

eellison commented:
Okay I added all of the functions in UnaryOps.cpp, BinaryOps.cpp, TensorCompare.cpp, TernaryOps.cpp, and I added a test that checks broadcasted shapes.  

I think it is pretty exhaustive. But I also want to note that we have landed other tags that are not actually being used and are incomplete (e.g. canonical). 

If people further thoughts on testing, I can do something more exhaustive that we agree on. One test that might make sense would be a DEBUG test that checks if when running TensorIterator that the most recently invoked dispatched operator includes the pointwise tag. This would be very expensive and we don't even have a DEBUG build right now but I think would work well.

ezyang commented:
> I think it is pretty exhaustive. But I also want to note that we have landed other tags that are not actually being used and are incomplete (e.g. canonical).

I agree canonical is problematic, but that's no excuse lol. (And pointwise is being used for much more substantive stuff in PyTorch than canonical right now, so let's hold it to a higher standard)

ezyang commented:
> If people further thoughts on testing, I can do something more exhaustive that we agree on. One test that might make sense would be a DEBUG test that checks if when running TensorIterator that the most recently invoked dispatched operator includes the pointwise tag. This would be very expensive and we don't even have a DEBUG build right now but I think would work well.

I suppose this raises the question; if we define a composite operator that only performs pointwise operations, should it have the pointwise tag? (This doesn't really matter for the partitioner as we assume all composite implicit operators have evaporated by now, but you could have a composite explicit that is defined in this way.)

I suggest we keep it simple: let's write a regex that extracts out the function list from UnaryOps.cpp, BinaryOps.cpp, TensorCompare.cpp, TernaryOps.cpp and then require that a function is marked pointwise if and only if it is in the extracted list. Done.

eellison commented:
> let's write a regex that extracts out the function list from UnaryOps.cpp, BinaryOps.cpp, TensorCompare.cpp, TernaryOps.cpp

This is done in the subsequent PRs

eellison committed:
Update on "Add Pointwise Tag from pointwise set in DTensor, use in aot_autograd partitioner"


Takes the pointwise op list from [DTensor](https://github.com/pytorch/pytorch/blob/master/torch/distributed/_tensor/ops/pointwise_ops.py#L36) as an initially starting point for pointwise ops, and feeds them to the aot autograd partitioner. 


[ghstack-poisoned]

ezyang commented:
> This is done in the subsequent PRs

feel free to squash

eellison committed:
Update on "Add Pointwise Tag from pointwise set in DTensor, use in aot_autograd partitioner"


Takes the pointwise op list from [DTensor](https://github.com/pytorch/pytorch/blob/master/torch/distributed/_tensor/ops/pointwise_ops.py#L36) as an initially starting point for pointwise ops, and feeds them to the aot autograd partitioner. 


[ghstack-poisoned]

ezyang commented:
This doesn't look like it does the regex?

ezyang commented:
I mean, have a unit test that runs whatever regex you did, and then check the tags match (both positively and negatively)

eellison committed:
Update on "Add Pointwise Tag from pointwise set in DTensor, use in aot_autograd partitioner"


Takes the pointwise op list from [DTensor](https://github.com/pytorch/pytorch/blob/master/torch/distributed/_tensor/ops/pointwise_ops.py#L36) as an initially starting point for pointwise ops, and feeds them to the aot autograd partitioner. 


[ghstack-poisoned]

ezyang reviewed (approved):
Good enough for now. Thank you for humoring me.

eellison committed:
Update on "Add Pointwise Tag from pointwise set in DTensor, use in aot_autograd partitioner"


Takes the pointwise op list from [DTensor](https://github.com/pytorch/pytorch/blob/master/torch/distributed/_tensor/ops/pointwise_ops.py#L36) as an initially starting point for pointwise ops, and feeds them to the aot autograd partitioner. 


[ghstack-poisoned]

eellison committed:
Update on "Add Pointwise Tag from pointwise set in DTensor, use in aot_autograd partitioner"


Takes the pointwise op list from [DTensor](https://github.com/pytorch/pytorch/blob/master/torch/distributed/_tensor/ops/pointwise_ops.py#L36) as an initially starting point for pointwise ops, and feeds them to the aot autograd partitioner. 


[ghstack-poisoned]

eellison commented:
@pytorchbot merge

bdhirsh commented on test/test_ops.py:
```diff
@@ -1973,6 +2060,59 @@ def map_to_fake(e):
+
+                    out_shape = torch._refs._broadcast_shapes(*shapes)
+
+                    for out_elem in tree_flatten(out):
```
do we actually have any pointwise ops that return more than 1 tensor?

bdhirsh replied:
oh maybe the `_foreach_*` ops fall in this category

eellison replied:
`frexp`

eellison commented:
@pytorchbot merge -f "unrelated flakey failure"
